### PR TITLE
Use general function to calculate o-o-b values and run failures

### PR DESCRIPTION
**Pull request by bpbond**
Step 1 of the filtering work in #64 — 

This PR adds a general `run_failures()` function in `workflow/02_pre_processing.Rmd` to calculate which observations are 'out' and then which runs 'fail'.

As far as I can see, it produces identical output to the `reviews` branch — as it should! — but @leeyap please check on your end!

Unknown

### CODE REVIEW

File: workflow/02_pre_processing.Rmd [hunk 1]
```diff
@@ -391,28 +380,26 @@ left_join(hector_temp_slice, run_fail_summary_tgav_ssp, by = c("run_number", "ss
...
 # Remove large data and force clear memory
 rm(trk_output)
-gc()
 
-trk_output_filtered_2 <- run_fail_summary_tgav %>% 
-  left_join(trk_output_filtered_1, by = c("ssp", "run_number")) %>% 
-  select(-fraction_fail)
+trk_output_filtered_2 <- tgav_h_hist %>% 
+  select(-fraction_out, -fail) %>% 
+  left_join(trk_output_filtered_1, by = "run_number")
```
Comment (bpbond) [inline workflow/02_pre_processing.Rmd:393]: @leeyap Note the ordering change here. Before, we joined and then dropped unneeded columns...but if they're not needed, and we're dealing with big datasets, it's better to drop them _before_ merging.

In other words, let's not make `left_join` do the work (in time and memory use) of creating them if they're just going to be removed.

**Review by leeyap — APPROVED** [review]
Much cleaner and working on my end - thanks!

**Comment by bpbond** [comment]
@leeyap Please take a look at https://rpubs.com/bpbond/989636

The "Filter: CMIP6 future" graph is obviously screwed up 🙃 and currently I've nuked your Figure 1 😞 (will get those back, promise)  ...

...but does section 0.4 (the summary of the filtering) make sense? I think we'll want something like this to report? In the code it's now super easy to add a new filter, but that required a lot of reworking in 9b78c6a.

**Comment by bpbond** [comment]
Woot! 

Screenshot 2023-01-06 at 6 50 56 AM

OK we need to look at and talk about the model's _performance_, but see in 14f7462 how easy it is now to add a new filtering step.

**Comment by leeyap** [comment]
Hm, I re-ran the workflow because I was having issues in #64 and am now erroring at this line:
```
fails_co2 <- run_failures("CMIP6 historical CO2",
                          model_output,
                          "CO2_concentration", 
                          cmip_bounds_h$year, 
                          cmip_bounds_h$minimum, 
                          cmip_bounds_h$maximum, 
                          var_ssp = NULL,
                          max_out_frac = MAX_FAILFRAC_HISTORICAL)

Error:
! All columns in a tibble must be vectors.
x Column `ssp` is NULL.
Backtrace:
  1. global run_failures(...)
  8. dplyr:::left_join.data.frame(., y, by = join_vars)
  9. dplyr:::join_mutate(...)
 11. tibble:::as_tibble.data.frame(y, .name_repair = "minimal")
 12. tibble:::lst_to_tibble(unclass(x), .rows, .name_repair)
 13. tibble:::check_valid_cols(x)
```

**Comment by bpbond** [comment]
Did you re-run the first script, running Hector, before trying to run `02_pre_processing.Rmd`? You'll need to do that.

**Comment by leeyap** [comment]
I did indeed, this is a fresh set of runs from `01_generate_data.Rmd`

**Comment by bpbond** [comment]
Hmm, sorry, not sure offhand but I will check later. Apologies that it didn't run right out of the box for you.

**Comment by leeyap** [comment]
Thanks, and no worries, could very well be an issue on my end.

**Comment by bpbond** [comment]
@leeyap I just cleared out the cache and re-ran everything (the 01 and 02 Rmarkdownfiles) without problem. Hmm. The latest commit in this branch is `14f7462` - that's where you're at?

**Comment by leeyap** [comment]
Yes! Still erroring, really not sure why

**Pull request merged**